morango/models.py

```diff
@@ -288,17 +288,33 @@ def _deserialize_store_model(self):
         klass_model = _profile_models[self.profile][self.model_name]
         # if store model marked as deleted, attempt to delete in app layer
         if self.deleted:
-            klass_model.objects.filter(id=self.id).delete()
+            # if hard deleted, propogate to related models
+            if self.hard_delete:
+                try:
+                    klass_model.objects.get(id=self.id).delete(hard_delete=True)
+                except klass_model.DoesNotExist:
+                    pass
+            else:
+                klass_model.objects.filter(id=self.id).delete()
         # inflate model and attempt to save
         else:
             app_model = klass_model.deserialize(json.loads(self.serialized))
             app_model._morango_source_id = self.source_id
             app_model._morango_partition = self.partition
             try:
+                app_model.clean_fields()
                 with mute_signals(signals.pre_save, signals.post_save):
                     app_model.save(update_dirty_bit_to=False)
             # if unable to save due to missing FKs, mark model as deleted
-            except ObjectDoesNotExist:
+            except ValidationError:
+                # check FKs in store to see if any of those models were hard deleted
+                fk_ids = [getattr(app_model, field.attname) for field in app_model._meta.fields if isinstance(field, ForeignKey)]
```

Comment:
I suppose it might be good to include OneToOneFields as well? Oh, nevermind, they're a subclass of ForeignKey! :fireworks: